src/emitter.js

```diff
@@ -0,0 +1,279 @@
+/**
+ * @license Copyright (c) 2003-2015, CKSource - Frederico Knabben. All rights reserved.
+ * For licensing, see LICENSE.md.
+ */
+
+'use strict';
+
+/**
+ * Mixin that injects the events API into its host.
+ *
+ * @class Emitter
+ * @static
+ */
+
+CKEDITOR.define( [ 'eventinfo', 'utils' ], function( EventInfo, utils ) {
+	var EmmitterMixin = {
+		/**
+		 * Registers a callback function to be executed when an event is fired.
+		 *
+		 * @param {String} event The name of the event.
+		 * @param {Function} callback The function to be called on event.
+		 * @param {Object} [ctx] The object that represents `this` in the callback. Defaults to the object firing the
+		 * event.
+		 * @param {Number} [priority] The priority of this callback in relation to other callbacks to that same event.
+		 * Lower values are called first. Defaults to `10`.
+		 */
+		on: function( event, callback, ctx, priority ) {
+			var callbacks = getCallbacks( this, event );
+
+			var wasAdded;
+
+			// Priority defaults to 10.
+			if ( typeof priority != 'number' ) {
+				priority = 10;
+			}
+
+			callback = {
+				callback: callback,
+				ctx: ctx || this,
+				priority: priority
+			};
+
+			// Add the callback to the list in the right priority position.
+			for ( var i = 0; i < callbacks.length ; i++ ) {
+				if ( callbacks[ i ].priority > priority ) {
+					callbacks.splice( i, 0, callback );
+					wasAdded = true;
+					break;
+				}
+			}
+
+			if ( !wasAdded ) {
+				callbacks.push( callback );
+			}
+		},
+
+		/**
+		 * Registers a callback function to be executed on the next time the event is fired only. This is similar to
+		 * calling `on()` followed by `off()` in the callback.
+		 *
+		 * @param {String} event The name of the event.
+		 * @param {Function} callback The function to be called on event.
+		 * @param {Object} [ctx] The object that represents `this` in the callback. Defaults to the object firing the
+		 * event.
+		 * @param {Number} [priority] The priority of this callback in relation to other callbacks to that same event.
+		 * Lower values are called first. Defaults to `10`.
+		 */
+		once: function( event, callback, ctx, priority ) {
+			var onceCallback = function( event ) {
+				// Go off() at the first call.
+				event.off();
+
+				// Go with the original callback.
+				callback.apply( this, arguments );
+			};
+
+			// Make the a similar on() call, simply replacing the callback.
+			this.on( event, onceCallback, ctx, priority );
+		},
+
+		/**
+		 * Stops executing the callback on the given event.
+		 *
+		 * @param {String} event The name of the event.
+		 * @param {Function} callback The function to stop being called.
+		 */
+		off: function( event, callback ) {
```

Comment:
What if we had the same callback but executed in different contexts? This would remove all those handlers. I think we need an optional context to be passed.